src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -230,125 +285,150 @@ RouterExecutorRun(QueryDesc *queryDesc, ScanDirection direction, long count, Tas
+		{
+			if ((affectedTupleCount == -1) ||
+				(affectedTupleCount == currentAffectedTupleCount))
+			{
+				affectedTupleCount = currentAffectedTupleCount;
+			}
+			else
+			{
+				ereport(WARNING,
+						(errmsg("modified "INT64_FORMAT " tuples, but expected "
+														"to modify "INT64_FORMAT,
+								currentAffectedTupleCount, affectedTupleCount),
+						 errdetail("modified placement on %s:%d",
+								   nodeName, nodePort)));
+			}
 
-	/* if all placements failed, error out */
-	if (list_length(failedPlacementList) == list_length(task->taskPlacementList))
-	{
-		ereport(ERROR, (errmsg("could not modify any active placements")));
+			resultsOK = true;
+
+			gotResults = true;
+
+			/*
+			 * Modifications have to be executed on all placements, but for
+			 * read queries we can stop here.
+			 */
+			if (!isModificationQuery)
+			{
+				break;
+			}
+		}
+		else
+		{
+			PurgeConnection(connection);
+
+			failedPlacementList = lappend(failedPlacementList, taskPlacement);
+
+			continue;
+		}
 	}
 
-	/* otherwise, mark failed placements as inactive: they're stale */
-	foreach(failedPlacementCell, failedPlacementList)
+	if (isModificationQuery)
 	{
-		ShardPlacement *failedPlacement = (ShardPlacement *) lfirst(failedPlacementCell);
-		uint64 shardLength = 0;
+		/* if all placements failed, error out */
+		if (list_length(failedPlacementList) == list_length(task->taskPlacementList))
+		{
+			ereport(ERROR, (errmsg("could not modify any active placements")));
+		}
+
+		/* otherwise, mark failed placements as inactive: they're stale */
+		foreach(failedPlacementCell, failedPlacementList)
+		{
+			ShardPlacement *failedPlacement =
+				(ShardPlacement *) lfirst(failedPlacementCell);
+			uint64 shardLength = 0;
+
+			DeleteShardPlacementRow(failedPlacement->shardId, failedPlacement->nodeName,
+									failedPlacement->nodePort);
+			InsertShardPlacementRow(failedPlacement->shardId, FILE_INACTIVE, shardLength,
+									failedPlacement->nodeName, failedPlacement->nodePort);
+		}
 
-		DeleteShardPlacementRow(failedPlacement->shardId, failedPlacement->nodeName,
-								failedPlacement->nodePort);
-		InsertShardPlacementRow(failedPlacement->shardId, FILE_INACTIVE, shardLength,
-								failedPlacement->nodeName, failedPlacement->nodePort);
+		executorState->es_processed = affectedTupleCount;
 	}
 
-	return affectedTupleCount;
+	return resultsOK;
 }
 
 
 /*
- * ExecuteSingleShardSelect executes, if not done already, the remote select query and
- * sends the resulting tuples to the given destination receiver. If the query fails on a
- * given placement, the function attempts it on its replica.
+ * ReturnRowsFromTuplestore moves rows from a given tuplestore into a
+ * receiver. It performs the necessary limiting to support cursors.
  */
-static void
-ExecuteSingleShardSelect(QueryDesc *queryDesc, uint64 tupleCount, Task *task,
-						 EState *executorState, TupleDesc tupleDescriptor,
-						 DestReceiver *destination)
+static int64
```

Comment:
Minor: It looks like we need to return `uint`. Also, this function does not use `executorState` parameter. You may want to remove that.